src/linear_algebra/pi_tensor_product.lean

```diff
@@ -0,0 +1,421 @@
+variables {s : ι → Type*} [∀ i, add_comm_monoid (s i)] [∀ i, semimodule R (s i)]
+variables {E : Type*} [add_comm_monoid E] [semimodule R E]
+variables {F : Type*} [add_comm_monoid F]
+
+namespace pi_tensor_product
+include R
+variables (R) (s)
+
+/-- The relation on `free_add_monoid (R × Π i, s i)` that generates a congruence whose quotient is
+the tensor product. -/
+inductive eqv : free_add_monoid (R × Π i, s i) → free_add_monoid (R × Π i, s i) → Prop
+| of_zero : ∀ (r : R) (f : Π i, s i) (i : ι) (hf : f i = 0), eqv (free_add_monoid.of (r, f)) 0
+| of_zero_scalar : ∀ (f : Π i, s i), eqv (free_add_monoid.of (0, f)) 0
+| of_add : ∀ (r : R) (f : Π i, s i) (i : ι) (m₁ m₂ : s i), eqv
+    (free_add_monoid.of (r, update f i m₁) + free_add_monoid.of (r, update f i m₂))
+    (free_add_monoid.of (r, update f i (m₁ + m₂)))
+| of_add_scalar : ∀ (r r' : R) (f : Π i, s i), eqv
+    (free_add_monoid.of (r, f) + free_add_monoid.of (r', f))
+    (free_add_monoid.of (r + r', f))
+| of_smul : ∀ (r : R) (f : Π i, s i) (i : ι) (r' : R), eqv
+    (free_add_monoid.of (r, update f i (r' • (f i))))
+    (free_add_monoid.of (r' * r, f))
+| add_comm : ∀ x y, eqv (x + y) (y + x)
+
+end pi_tensor_product
+
+variables (R) (s)
+
+/-- `pi_tensor_product R s` with `R` a commutative semiring and `s : ι → Type*` is the tensor
+  product of all the `s i`'s. This is denoted by `⨂[R] i, s i`. -/
+def pi_tensor_product : Type* :=
+(add_con_gen (pi_tensor_product.eqv R s)).quotient
+
+variables {R}
+
+/- This enables the notation `⨂[R] i : ι, s i` for the pi tensor product, given `s : ι → Type*`. -/
+localized "notation `⨂[`:100 R `] ` binders `, ` r:(scoped:67 f, pi_tensor_product R f) := r"
+  in tensor_product
+
+open_locale tensor_product
+
+namespace pi_tensor_product
+
+section module
+
+instance : add_comm_monoid (⨂[R] i, s i) :=
+{ add_comm := λ x y, add_con.induction_on₂ x y $ λ x y, quotient.sound' $
+    add_con_gen.rel.of _ _ $ eqv.add_comm _ _,
+  .. (add_con_gen (pi_tensor_product.eqv R s)).add_monoid }
+
+instance : inhabited (⨂[R] i, s i) := ⟨0⟩
+
+variables (R) {s}
+
+/-- `tprod_coeff R r f` with `r : R` and `f : Π i, s i` is the tensor product of the vectors `f i`
+over all `i : ι`, multiplied by the coefficient `r`. -/
+def tprod_coeff (r : R) (f : Π i, s i) : ⨂[R] i, s i := add_con.mk' _ $ free_add_monoid.of (r, f)
+
+/-- `tprod R f` with `f : Π i, s i` is the tensor product of the vectors `f i` over all `i : ι`. -/
+def tprod (f : Π i, s i) : ⨂[R] i, s i := tprod_coeff R 1 f
+
+variables {R}
+
+notation `⨂ₜ[`:100 R`] ` binders `, ` r:(scoped:67 f, tprod R f) := r
+
+lemma zero_tprod_coeff (f : Π i, s i) : tprod_coeff R 0 f = 0 :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_zero_scalar _
+
+lemma zero_tprod_coeff' (z : R) (f : Π i, s i) (i : ι) (hf: f i = 0) : tprod_coeff R z f = 0 :=
+quotient.sound' $ add_con_gen.rel.of _ _ $ eqv.of_zero _ _ i hf
+
+lemma zero_tprod (f : Π i, s i) (i : ι) (hf: f i = 0) : tprod R f = 0 :=
```

Comment:
Since you have the notation, can you use it here?

```suggestion
lemma zero_tprod (f : Π i, s i) (i : ι) (hf: f i = 0) : ⨂ₜ[R] i, f i = 0 :=
```

Reply:
Right -- I'm having second thoughts about this notation. I initially put it in thinking it would be nice, but when I started using it the extra index just made things look cluttered. I'm not sure if I should just remove it entirely though.